File: algo-library/python/hash/ZobristHash.py
from random import randint
class ZobristHash:
  # 比較したい配列を引数に渡す
  def __init__(self,A: list, B: list) -> None:
    MAX = (1<<61)-1
    self.hash = dict()
    for a in A: self.hash[a] = randint(1,MAX)
    for b in B: self.hash[b] = randint(1,MAX)
    # ハッシュを取る
    self.hashA,self.hashB = [0],[0]
    setA,setB = set(),set()
    for a in A: 
      if a in setA:
        self.hashA.append(self.hashA[-1])
      else:
        self.hashA.append(self.hashA[-1] ^ self.hash[a])
        setA.add(a)
    for b in B: 
      if b in setB:
        self.hashB.append(self.hashB[-1])
      else:
        self.hashB.append(self.hashB[-1] ^ self.hash[b])
        setB.add(b)
    
  # 1-index(A_1, A_2, ...)
  # (A_al,A_{al+1},...,A_ar) == (B_al,B_{al+1},...,B_ar) を返す
  def isEqual(self,al,ar,bl,br):
    return (self.hashA[ar] - self.hashA[al-1]) == (self.hashB[br] - self.hashB[bl-1])

File: algorithm_library/cpp-from-py/hash/ZobristHash.cpp
#include <bits/stdc++.h>
using namespace std;

template <class T>
struct ZobristHash {
    static constexpr unsigned long long MAXV = (1ULL << 61) - 1;
    unordered_map<T, unsigned long long> hash_value;
    vector<unsigned long long> hashA, hashB;
    mt19937_64 rng;

    ZobristHash(const vector<T>& A, const vector<T>& B) : rng(chrono::steady_clock::now().time_since_epoch().count()) {
        for (const auto& a : A) hash_value[a] = randint();
        for (const auto& b : B) hash_value[b] = randint();
        hashA = {0};
        hashB = {0};
        unordered_set<T> setA, setB;
        for (const auto& a : A) {
            if (setA.count(a)) hashA.push_back(hashA.back());
            else {
                hashA.push_back(hashA.back() ^ hash_value[a]);
                setA.insert(a);
            }
        }
        for (const auto& b : B) {
            if (setB.count(b)) hashB.push_back(hashB.back());
            else {
                hashB.push_back(hashB.back() ^ hash_value[b]);
                setB.insert(b);
            }
        }
    }

    unsigned long long randint() {
        return uniform_int_distribution<unsigned long long>(1, MAXV)(rng);
    }

    bool isEqual(int al, int ar, int bl, int br) const {
        return (hashA[ar] ^ hashA[al - 1]) == (hashB[br] ^ hashB[bl - 1]);
    }
};
